### PR TITLE
fix(functions): prevent etag salt truncation when containing hyphens

### PR DESCRIPTION
**Pull request by avolkovi**
### Description
Fixes an issue where Cloud Functions declarative security role etags were incorrectly truncated if the salt contained a hyphen (which occurs with ~21.3% probability). Invariant suffix-aware slicing strips the 33-character -<hash> suffix, preserving standard 10-character salts and legacy pre-release salts, while falling back gracefully for short test mocks.

Fixes http://b/557361863

### Scenarios Tested
- Verified round-trip salt extraction and etag recomputation across random salts and edge cases (single/multi-hyphen, underscores).
- Verified fast-path equality branch without triggering IAM permission check when roles are unchanged.
- Verified fast-path equality branch for legacy pre-release truncated salts (< 10 chars).
- Verified permission check triggers when required roles change while still preserving salt.
- Verified test mock fallbacks (<= 33 characters).

### Sample Commands
`firebase deploy --only functions`

### CODE REVIEW

File: src/gcp/iam.ts [hunk 1]
```diff
@@ -329,24 +329,18 @@ export async function generateManagedServiceAccountName(
...
-  }
   const sorted = Array.from(roles).sort();
   const hashBuffer = crypto
     .createHash("sha256")
-    .update(salt + sorted.join(","))
+    .update(sorted.join(","))
     .digest();
   let hashStr = "";
   for (const byte of hashBuffer) {
     hashStr += BASE38_CHARS[byte % 38];
```
Comment (ajperel) [inline src/gcp/iam.ts]: label etag can't start with -_ and maybe not a number. that's why the salt computation started with: `salt = String.fromCharCode(97 + Math.floor(Math.random() * 26)); // a-z`

We need to also guarantee this for the first character of the hashStr without the salt.

**Comment by ajperel** [comment]
Done in https://github.com/firebase/firebase-tools/pull/11080
